src/main/java/com/cleanroommc/modularui/widgets/DropDownMenu.java

```diff
@@ -0,0 +1,244 @@
+        setArrows(GuiTextures.ARROW_UP, GuiTextures.ARROW_DOWN);
+    }
+
+    public int getSelectedIndex() {
+        return menu.getCurrentIndex();
+    }
+
+    public DropDownMenu setSelectedIndex(int index) {
+        menu.setCurrentIndex(index);
+        return getThis();
+    }
+
+    public DropDownMenu addChoice(Function<Integer, DropDownItem> itemGetter) {
+        menu.addChoice(itemGetter);
+        return getThis();
+    }
+
+    public DropDownMenu setArrows(IDrawable arrowClosed, IDrawable arrowOpened) {
+        this.arrowClosed = arrowClosed;
+        this.arrowOpened = arrowOpened;
+        return getThis();
+    }
+
+    public DropDownMenu setMaxItemsToDisplay(int maxItems) {
+        menu.setMaxItemsToDisplay(maxItems);
+        return getThis();
+    }
+
+    public DropDownMenu addChoice(ItemSelected onSelect, IDrawable... drawable) {
+        DropDownItem button = new DropDownItem();
+        return addChoice(index ->
+            button.onMouseReleased(m -> {
+                menu.setOpened(false);
+                menu.setCurrentIndex(index);
+                onSelect.selected(this);
+                return true;
+                })
+            .overlay(drawable));
+    }
+
+    public DropDownMenu addChoice(ItemSelected onSelect, String text) {
+        return addChoice(onSelect, new StringKey(text));
+    }
+
+    public DropDownMenu setDropDownDirection(DropDownDirection direction) {
+        menu.setDropDownDirection(direction);
+        return getThis();
+    }
+
+    @Override
+    public @NotNull Result onMousePressed(int mouseButton) {
+        if (!menu.isOpen()) {
+            menu.setOpened(true);
+            menu.setEnabled(true);
+            return Result.SUCCESS;
+        }
+        menu.setOpened(false);
+        menu.setEnabled(false);
+        return Result.SUCCESS;
+    }
+
+    @Override
+    public void draw(GuiContext context) {
+        super.draw(context);
+        Area area = getArea();
+        int smallerSide = Math.min(area.width, area.height);
+        if (menu.getSelectedItem() != null) {
+            menu.getSelectedItem().setEnabled(true);
+            menu.getSelectedItem().drawBackground(context);
+            menu.getSelectedItem().draw(context);
+            menu.getSelectedItem().drawForeground(context);
+        } else {
+            NONE.draw(context, 0, 0, area.width - smallerSide, area.height);
+        }
+
+        if (menu.isOpen()) {
+            arrowOpened.draw(context, area.width - smallerSide , 0, smallerSide, smallerSide);
+        } else {
+            arrowClosed.draw(context, area.width - smallerSide , 0, smallerSide, smallerSide);
+        }
+    }
+
+    @Override
+    public DropDownMenu background(IDrawable... background) {
+        menu.background(background);
+        return super.background(background);
+    }
+
+    public enum DropDownDirection {
+        UP(0, -1),
+        DOWN(0, 1);
+
+        private int xOffset = 0;
+        private int yOffset = 0;
```

Comment:
nit: redundant assignments

Reply:
should I also make those final afterwards?

Reply:
yeah